zkapp/contract.ts

```diff
@@ -51,21 +57,29 @@ export class HelloWorld extends SmartContract {
     this.emitEvent('singleField', x);
   }
 
-  @method async emitStructEvent() {
-    const x = this.x.getAndRequireEquals();
-    const y = this.y.getAndRequireEquals();
-    const z = this.z.getAndRequireEquals();
-    this.emitEvent(
-      'struct',
-      new TestStruct({ x, y, z, address: this.address })
-    );
+  @method async emitStructEvent(struct: TestStruct) {
+    this.emitEvent('struct', struct);
+  }
+
+  @method async emitStructsEvent(structs: TestStructArray) {
+    this.emitEvent('structs', structs);
   }
 
-  @method async emitStructAction() {
+  /**
+   * This method always emits the same action.
+   * It has limited utility in generating realistic test cases.
+   * We should use more dynamic methods to generate actions.
+   */
+  @method async emitStaticStructAction() {
     const x = this.x.getAndRequireEquals();
     const y = this.y.getAndRequireEquals();
     const z = this.z.getAndRequireEquals();
-    this.reducer.dispatch(new TestStruct({ x, y, z, address: this.address }));
+    const struct = new TestStruct({ x, y, z, address: this.address });
+    this.reducer.dispatch({ structs: [struct, struct, struct] });
```

Comment:
One contract may have many types of events, but only one type of action.  To keep the tests slightly simple and fast, I updated this action type to the most complex: One action with multiple field arrays with multiple fields.

The alternative was writing a new contract where one has a simple action type and one has a complex action type, and we would need to await compile, deploy, etc... on both.